File: android/app/src/main/java/com/facecolour/app/engine/Product.kt
package com.facecolour.app.engine

import kotlin.math.abs

enum class ProductCategory { MAKEUP, FOUNDATION, CLOTHING, ACCESSORY, OTHER }

/** Shoppable product. `productUrl` is the retailer deep-link (no in-app checkout). */
data class Product(
    val id: String,
    val title: String,
    val brand: String?,
    val price: String?,
    val imageUrl: String?,
    val productUrl: String,
    val colorHex: String?,
    val category: ProductCategory?,
    val seasons: List<Season>?,
    val monkTone: Int?
)

/** Pure relevance filtering — mirrors iOS `ProductMatcher`. */
object ProductMatcher {
    fun filter(
        products: List<Product>,
        season: Season?,
        monkTone: Int?,
        toneTolerance: Int = 1
    ): List<Product> = products.filter { p ->
        val seasonOk = if (season != null && !p.seasons.isNullOrEmpty()) p.seasons.contains(season) else true
        val toneOk = if (monkTone != null && p.monkTone != null) abs(p.monkTone - monkTone) <= toneTolerance else true
        seasonOk && toneOk
    }
}

File: Sources/FaceColour/ColorEngine/Product.swift
import Foundation

enum ProductCategory: String, Codable, Equatable {
    case makeup, foundation, clothing, accessory, other
}

/// A shoppable product. `productURL` is the retailer deep-link (no in-app checkout).
/// `seasons` / `monkTone` are optional relevance tags used for filtering.
struct Product: Codable, Equatable, Identifiable {
    let id: String
    let title: String
    let brand: String?
    let price: String?
    let imageURL: URL?
    let productURL: URL
    let colorHex: String?
    let category: ProductCategory?
    let seasons: [Season]?
    let monkTone: Int?
}

/// Pure relevance filtering — shared with the Android port.
enum ProductMatcher {
    /// Keep products that suit the season (untagged = general) and, for
    /// shade-specific products, are within `toneTolerance` of the matched tone.
    static func filter(_ products: [Product],
                       season: Season?,
                       monkTone: Int?,
                       toneTolerance: Int = 1) -> [Product] {
        products.filter { product in
            let seasonOK: Bool
            if let season, let tags = product.seasons, !tags.isEmpty {
                seasonOK = tags.contains(season)
            } else {
                seasonOK = true
            }

            let toneOK: Bool
            if let monkTone, let tone = product.monkTone {
                toneOK = abs(tone - monkTone) <= toneTolerance
            } else {
                toneOK = true
            }

            return seasonOK && toneOK
        }
    }
}
